Rewrite OpenXR hand tracking to reflect new architecture (#9499)
godotengine/godot-docs: tutorials/xr/openxr_hand_tracking.rst

`here <https://github.com/godotengine/godot-demo-projects/tree/master/xr/openxr_hand_tracking_demo>`_.

The Hand Tracking API
---------------------

As mentioned in our introduction, the hand tracking API is primarily used with optical hand tracking
and on many platforms only works when the user is not holding a controller.
Some platforms support controller inferred hand tracking meaning that you will get hand tracking data
even if the user is holding a controller.
This includes SteamVR, Meta Quest (currently native only but Meta link support is likely coming)
and hopefully soon others as well.

The hand tracking implementation in Godot has been standardised around the Godot Humanoid Skeleton
and works both in OpenXR and WebXR. The instructions below will thus work in both environments.

In order to use the hand tracking API with OpenXR you first need to enable it.
This can be done in the project settings:

**add image of project settings**

For some standalone XR devices you also need to configure the hand tracking extension in export settings,
for instance for Meta Quest:

.. image:: img/xr_enable_handtracking.webp

> **replace this image with an updated one**

Author: Reminder to self, need to do this.

.. _doc_openxr_hand_tracking:

OpenXR hand tracking
========================

Introduction
------------

.. note::

    This page focuses specifically on the feature set exposed through OpenXR.
    Parts of the functionality presented here also applies to WebXR and can by provided
    by other XR interfaces.

When discussing hand tracking it is important to know that there are differences of opinion as to where lines are drawn.
The practical result of this is that there are differences in implementation between the different OpenXR runtimes.

> You may find yourself in a place where chosen hardware doesn't support a piece of the puzzle or does things different

Reviewer: ```suggestion
You may find yourself in a place where chosen hardware doesn't support a piece of the puzzle or does things differently
```

enough from the other platforms that you need to do extra work.

That said, recent improvements to the OpenXR specification are closing these gaps and as platforms implement
these improvements we are getting closer to a future where we have either full portability between platforms
or at least a clear way to detect the capabilities of a platform.

When we look at the early days of VR the focus of the major platforms was on tracked controller based input.
Here we are tracking a physical device that also has buttons for further input.

> From the tracking data we can infer the location of the players hands but no further information is known,
> traditionally it was left up to the game to implement a mechanism to display the players hand and animate

Reviewer: ```suggestion
From the tracking data we can infer the location of the player's hands but no further information is known,
traditionally it was left up to the game to implement a mechanism to display the player's hand and animate
```

the fingers based on further input from the controller, be it due to buttons being pressed or through proximity
sensors.
Often fingers are also placed based on context, what is the user holding, what is the action a user is performing.

> More recently optical hand tracking has become a popular solution, where cameras track the users hands

Reviewer: ```suggestion
More recently optical hand tracking has become a popular solution, where cameras track the user's hands
```

> and full tracked data for the hand and finger position became available.

Reviewer 2: ```suggestion
and full tracking data for the hand and finger positions becomes available.
```

Many vendors saw this as completely separate from controller tracking and introduced independent APIs to

> access hand and finger position and orientation data.

Reviewer 2: ```suggestion
access hand and finger positions and orientation data.
```

> When handling input it was up to the game developer to implement a gesture detection mechanism.

Reviewer 2: ```suggestion
When handling input, it was up to the game developer to implement a gesture detection mechanism.
```

while optical hand tracking is primarily handled by the hand tracking API extension.

However, the world is not that black and white and we're seeing a number of scenarios
where we cross the line:

 *  Devices that fit in both categories, such as tracked gloves
    and controllers such as the Index controller that also perform finger tracking.
 *  XR Runtimes that implement infered hand tracking from controller data as a means
    to solve proper finger placement for multiple controllers.
 *  XR applications that wish to seamlessly switch between controller and hand tracking
    offering the same user experience regardless of approach used.

OpenXR is answering this call by introducing further extensions that lets us query the capabilities of
the XR runtime/hardware or that add further functionality accross this divide.
The problem that currently does remain is that there are gaps in adopting these extensions,
with some platforms thus not reporting capabilities to their full extent.
As such you may need to test for the features available on specific hardware
and adjust your approach accordingly.

Demo project
------------

> The information presented on this page was used to create a demo projection that can be found

Reviewer 2: ```suggestion
The information presented on this page was used to create a demo project that can be found
```

`here <https://github.com/godotengine/godot-demo-projects/tree/master/xr/openxr_hand_tracking_demo>`_.

The Hand Tracking API
---------------------

As mentioned in our introduction, the hand tracking API is primarily used with optical hand tracking
and on many platforms only works when the user is not holding a controller.
Some platforms support controller inferred hand tracking meaning that you will get hand tracking data
even if the user is holding a controller.
This includes SteamVR, Meta Quest (currently native only but Meta link support is likely coming)
and hopefully soon others as well.

> The hand tracking implementation in Godot has been standardised around the Godot Humanoid Skeleton

Reviewer 2: Probably not important, but I think we try to use US (vs UK) spellings:

```suggestion
The hand tracking implementation in Godot has been standardized around the Godot Humanoid Skeleton
```

and works both in OpenXR and WebXR. The instructions below will thus work in both environments.

In order to use the hand tracking API with OpenXR you first need to enable it.
This can be done in the project settings:

.. image:: img/xr_enable_handtracking.webp

For some standalone XR devices you also need to configure the hand tracking extension in export settings,
for instance for Meta Quest:

.. image:: img/openxr_enable_hand_tracking_meta.webp

Now you need to add 3 components into your scene for each hand:

 *  A tracked node to position the hand.
 *  A properly skinned hand mesh with skeleton.
 *  A skeleton modifier that applies finger tracking data to the skeleton.

.. image:: img/openxr_hand_tracking_nodes.webp

Hand tracking node
^^^^^^^^^^^^^^^^^^

> The hand tracking system uses separate hand trackers to track the position of the players hands

Reviewer: ```suggestion
The hand tracking system uses separate hand trackers to track the position of the player's hands
```

within our tracking space. 

This information has been separated out for the following use cases:

 *  Tracking happens in the local space of the :ref:`XROrigin3D <class_xrorigin3d>` node.
    This node must be a child of the `XROrigin3D` node in order to be correctly placed.
 *  This node can be used as an IK target when an upper body mesh with arms is used instead
    of separate hand meshes.
 *  Actual placement of the hands may be loosly bound to the tracking in scenarios such as

>     avatar creation UIs, fake mirrors or similar situations

Reviewer: ```suggestion
    avatar creation UIs, fake mirrors, or similar situations
```

>     resulting in the hand mesh and finger tracking being localised elsewhere.

Reviewer 2: Another US vs UK spelling:

```suggestion
    resulting in the hand mesh and finger tracking being localized elsewhere.
```

You can access this information through code:

.. code-block:: gdscript

    var hand_tracker : XRHandTracker = XRServer.get_tracker('/user/hand_tracker/left')
    if hand_tracker:
        if hand_tracker.has_tracking_data:
            if hand_tracker.hand_tracking_source == XRHandTracker.HAND_TRACKING_SOURCE_UNKNOWN:
                print("Hand tracking source unknown")
            elif hand_tracker.hand_tracking_source == XRHandTracker.HAND_TRACKING_SOURCE_UNOBSTRUCTED:
                print("Hand tracking source is optical hand tracking")
            elif hand_tracker.hand_tracking_source == XRHandTracker.HAND_TRACKING_SOURCE_CONTROLLER:
                print("Hand tracking data is inferred from controller data")
            else:
                print("Unknown hand tracking source ", hand_tracker.hand_tracking_source)
        else:
            print("Hand is currently not being tracked")
    else:
        print("No hand tracker registered")

This example simply logs the state for the left hand.

If in this example no hand tracker is returned by ``get_tracker``,
this means the hand tracking API is not supported on the XR runtime at all.

> If there is a tracker but `has_tracking_data` is false, the users hand is currently not being tracked.

Reviewer: ```suggestion
If there is a tracker but `has_tracking_data` is false, the user's hand is currently not being tracked.
```

This is likely caused by one of the following reasons:

 *  The player's hand is not visible by any of the tracking cameras on the headset
 *  The player is currently using a controller and the headset only supports optical hand tracking
 *  The controller is turned off and only controller hand tracking is supported.

Handling user input
-------------------

Reacting to actions performed by the user is handled through :ref:`doc_xr_action_map`
if controllers are used.
In the action map you can map various inputs like the trigger or joystick on the controller
to an action. This can then drive logic in your game.

When hand tracking is used we originally had no such inputs,
inputs are driven by gestures made by the user such as making a fist to grab
or pinching the thumb and index finger together to select something.
It was up to the game developer to implement this.

> Recognising that there is an increasing demand for applications that can switch seamlessly

Reviewer 2: US vs UK spelling:

```suggestion
Recognizing that there is an increasing demand for applications that can switch seamlessly
```

between controller and hand tracking and the need some form of basic input capability,
a number of extensions were added to the specification that provide some basic gesture recognision
and can be used with the action map.

The hand interaction profile
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The `hand interaction profile extension <https://github.khronos.org/OpenXR-Inventory/extension_support.html#XR_EXT_hand_interaction>`_

> is a new core extension which supports pinch, grasp and poke gestures and related poses.

Reviewer: ```suggestion
is a new core extension which supports pinch, grasp, and poke gestures and related poses.
```

See the Microsoft hand interaction profile for the gesture support.

The defining difference is that this extension introduces two new trackers,
``/user/hand_htc/left`` and ``/user/hand_htc/right``.
This means that extra logic needs to be implemented to switch between the default trackers
and the HTC specific trackers when the user puts down, or picks up, their controller.

Simple controller profile
^^^^^^^^^^^^^^^^^^^^^^^^^

The simple controller profile is a standard core profile defined as a fallback profile
when a controller is used for which no profile exists.

There are a number of OpenXR runtimes that will mimic controllers through
the simple controller profile when hand tracking is used.

Unfortunately there is no sound way to determine whether an unknown controller is used
or whether hand tracking is emulating a controller through this profile.

.. image:: img/openxr_simple_controller_hand.webp

XR runtimes are free to define how the simple controller profile operates,
so there is also no certainty to how this profile is mapped to gestures.

The most common mapping seems to be that ``select click`` is true
when the tip of the thumb and index fingers are touching while the

> users palm is facing away from the user.

Reviewer: ```suggestion
user's palm is facing away from the user.
```

``menu click`` will be true when tip of the thumb and index fingers

> are touching while the users palm is facing towards the user.

Reviewer: ```suggestion
are touching while the user's palm is facing towards the user.
```
